### PR TITLE
Implemented a smarter TokenStream concatenation system

### PR DESCRIPTION
**Pull request by cgswords**
The new algorithm performs 'aggressive compacting' during concatenation as follows:
- If the nodes' combined total total length is less than 32, we copy both of
  them into a new vector and build a new leaf node.
- If one node is an internal node and the other is a 'small' leaf (length<32),
  we recur down the internal node on the appropriate side.
  - Otherwise, we construct a new internal node that points to them as left and
    right.

This should produce notably better behavior than the current concatenation implementation.

### CODE REVIEW

File: src/libsyntax/tokenstream.rs [hunk 1]
```diff
@@ -496,14 +526,17 @@ impl InternalTS {
...
     pub fn mk_empty() -> TokenStream {
         TokenStream { ts: InternalTS::Empty(DUMMY_SP) }
     }
 
+    // Construct an empty node with the provided span.
     fn mk_spanned_empty(sp: Span) -> TokenStream {
         TokenStream { ts: InternalTS::Empty(sp) }
     }
 
+    // Construct an leaf node with a 0 offset and length equivalent to the input.
```
Comment (nrc) [inline src/libsyntax/tokenstream.rs]: Typo: here and below s/an/a

File: src/libsyntax/tokenstream.rs [hunk 2]
```diff
@@ -516,6 +549,7 @@ impl TokenStream {
         }
     }
 
+    // Construct an leaft node with the provided values.
```
Comment (nrc) [inline src/libsyntax/tokenstream.rs]: typo: leaft

File: src/libsyntax/tokenstream.rs [hunk 3]
```diff
@@ -483,6 +483,36 @@ impl InternalTS {
...
+            match *ts {
+                InternalTS::Empty(..) => Vec::new(),
+                InternalTS::Leaf { ref tts, offset, len, .. } => {
+                    tts[offset..offset + len].iter().collect()
+                }
+                InternalTS::Node { ref left, ref right, .. } => {
+                    let mut v1 = internal_to_vec(left);
+                    let mut v2 = internal_to_vec(right);
+                    v1.append(&mut v2);
+                    v1
```
Comment (nrc) [inline src/libsyntax/tokenstream.rs]: I expect it is more efficient to allocate a single (mut) vector with capacity == length of the TS, then pass it to the recursive calls, appending as you go down.

File: src/libsyntax/tokenstream.rs [hunk 4]
```diff
@@ -561,11 +596,56 @@ impl TokenStream {
...
+        // - If one node is an internal node and the other is a 'small' leaf (length<32),
+        //   we recur down the internal node on the appropriate side.
+        // - Otherwise, we construct a new internal node that points to them as left and
+        // right.
+        fn concat_internal(left: Rc<InternalTS>, right: Rc<InternalTS>) -> TokenStream {
+            let llen = left.len();
+            let rlen = right.len();
+            let len = llen + rlen;
+            let span = combine_spans(left.span(), right.span());
+            if len < 32 {
```
Comment (nrc) [inline src/libsyntax/tokenstream.rs]: Should these `<` be `<=`? I think we should allow leaf nodes which are 32 long. It's probably also a good idea to pull out `32` into a `const`.

**Comment by nrc** [comment]
Some pretty minor comments inline, otherwise looks good.

File: src/libsyntax/tokenstream.rs [hunk 5]
```diff
@@ -340,6 +340,8 @@ pub struct TokenStream {
     ts: InternalTS,
 }
 
+static LEAF_SIZE : usize = 32;
```
Comment (nrc) [inline src/libsyntax/tokenstream.rs]: should be const, not static, also no space before the colon

Comment (nrc) [inline src/libsyntax/tokenstream.rs]: and could you add a comment please?

**Comment by nrc** [comment]
r=me with those last two things addressed

**Comment by cgswords** [comment]
Addressed.

**Pull request merged**